nilearn/regions/parcellations.py

```diff
@@ -409,7 +410,11 @@ def _raw_fit(self, data):
             )
 
         else:
-            mask_ = safe_get_data(mask_img_).astype(bool)
+            if isinstance(mask_img_, SurfaceImage):
+                # concatenate arrays stored in both parts of the mask data
+                mask_ = np.concatenate(mask_img_.data.parts.values())
+            else:
+                mask_ = safe_get_data(mask_img_).astype(bool)
             shape = mask_.shape
             connectivity = image.grid_to_graph(
                 n_x=shape[0], n_y=shape[1], n_z=shape[2], mask=mask_
```

Comment:
Now [_BaseDecomposition](https://github.com/nilearn/nilearn/blob/135e5388386c24ca879328040273127f8acb4a6a/nilearn/decomposition/_base.py#L247) class is accepting SurfaceMasker almost everywhere, up to the point where we need `SurfaceMapsMasker`.

But as @bthirion suggested here: https://github.com/nilearn/nilearn/pull/4577#discussion_r1843821348, allowing wards clustering (or agglomerative clustering) would require more work.

Reply:
It's just calculating the edges between the nodes (that are voxels here), i.e., the adjacency matrix in a 3D volume.

IIUC, I think we could reuse what was done in rena clustering to calculate this connectivity/adjacency matrix for surfaces:
https://github.com/nilearn/nilearn/blob/4bcd033d30ca42d32ecdc408b24d3928112cea12/nilearn/regions/rena_clustering.py#L283

Reply:
done!